hub/api/dataset.py

```diff
@@ -1,33 +1,63 @@
+from hub.constants import META_FILENAME
 from hub.api.tensor import Tensor
 from hub.util.slice import merge_slices
+from hub.util.path import local_provider_from_path
 from hub.util.exceptions import (
     TensorNotFoundError,
     InvalidKeyTypeError,
     UnsupportedTensorTypeError,
 )
-from typing import Union, Dict
+from hub.core.typing import StorageProvider
+from hub.core.storage import MemoryProvider
+from hub.core.chunk_engine.read import read_dataset_meta, read_tensor_meta
+from hub.core.chunk_engine.write import write_array, write_dataset_meta
+from typing import Union, Dict, Optional
 import numpy as np
+import os
 
 
 class Dataset:
-    def __init__(self, path: str, mode: str = "a", ds_slice: slice = slice(None)):
+    def __init__(
+        self,
+        path: str,
+        mode: str = "a",
+        ds_slice: slice = slice(None),
+        provider: Optional[StorageProvider] = None,
+    ):
         """Initialize a new or existing dataset.
 
         Args:
             path (str): The location of the dataset.
                 Can be a local path, or a url to a cloud storage provider.
-            mode (str, optional): Mode in which the dataset is opened.
+            mode (str): Mode in which the dataset is opened.
                 Supported modes include ("r", "w", "a") plus an optional "+" suffix.
                 Defaults to "a".
-            ds_slice (slice, optional): The slice object restricting the view
+            ds_slice (slice): The slice object restricting the view
                 of this dataset's tensors. Defaults to slice(None, None, None).
                 Used internally for iteration.
+            provider (StorageProvider, optional): The storage provider used to access
+                the data stored by this dataset. Will not be used if given path is valid.
+
+        Raises:
+            ValueError: If an existing local path is given, it must be a directory.
         """
         self.path = path
         self.mode = mode
         self.slice = ds_slice
-        # TODO: read metadata and initialize tensors
+
+        self.provider = local_provider_from_path(path)
+        if self.provider is None:
```

Comment:
another reason to create `provider_from_path` instead of just `local_provider_from_path`: this code will be a one-liner here and if we need to use this util later we can